src/rdkafka_broker.c

```diff
@@ -407,8 +407,12 @@ void rd_kafka_broker_set_state(rd_kafka_broker_t *rkb, int state) {
                                  * broker connections. */
                                 rd_atomic32_add(&rkb->rkb_rk->rk_broker_up_cnt,
                                                 1);
+                                rd_kafka_broker_unlock(rkb);
```

Comment:
I think there should be only one thread setting state at any given amount of time. Unlocking in the middle might introduce the path which is not expected. For example, 

Thread A: set state and releases lock
Thread B: acquires the lock again. Doesn't enter this condition and unlocks after the function exits.
Thread A: acquires the lock again.

The state would be inconsistent in the scenario. I would avoid unlocking inside this function.

Reply:
There are many fields getting changed in this function. If we unlock in between and then try to acquire the lock again, we   will not be certain if any other thread would have changed the value or not. I think we should avoid unlocking here in between the function. Can we move this reset part to another function or at the end of the function somehow?